diagrams/HighLevelStateDiagram.md

```diff
@@ -0,0 +1,131 @@
+
+//All states
+```plantuml
+@startuml
+hide empty description
+
+[*] --> FollowerState
+state LeaderElected <<fork>>
+FollowerState --> LeaderElection
+LeaderElection --> LeaderElected
+
+LeaderElected --> HeartBeat
+LeaderElected --> LogReplication
+
+LogReplication --> HeartBeat
+HeartBeat --> [*]
+
+@enduml
+
+```
+
+//Follower state
+```plantuml
+
+@startuml
+hide empty description
+
+[*] --> FollowerState
+
+state FollowerState {
+ [*] --> Node1
+ [*] --> Node2
+ [*] --> Node3
+
+}
+ Node2 --> Candidate
+ Candidate : Node2 reached timeout. \nBecome a candidate
+
+ Candidate --> [*] : End of start state. \nLeader election beginning
+
+@enduml
+
+```
+
+
+//Leader election
+```plantuml
+
+@startuml
+hide empty description
+
+[*] --> Candidate
+[*] --> Node1
+[*] --> Node3
+
+state Candidate
+Candidate : Node2 reached timeout.\nBecome a candidate
+
+state LeaderElection {
+ state SendVotes
+ state ReceiveVotes
+
+
+}
+Candidate --> SendVotes : Start of election
+SendVotes --> Node1 : Send vote request
+SendVotes --> Node3 : Send vote request
+
+Node1 --> ReceiveVotes : Reply with vote
+Node3 --> ReceiveVotes : Reply with vote
+
+state Leader : Election ended.\nCandidate became a leader.
+ReceiveVotes --> Leader
+
+Leader --> [*]
+@enduml
+
+```
+
+
+//Log replication
+```plantuml
+
+@startuml
+hide empty description
```

Comment:
Please, describe the states of only the leader in this case and model behaviour of other nodes as signals. If the diagram become complex, make a simple node and separate diagram that will flesh out details, what is happening inside, eg https://gist.github.com/jnick26/11868e49a2e713fa5964804e36de18c3 become hairy very quickly.


referr to 5.3 and 5.4.2 from https://raft.github.io/raft.pdf